src/algebra/module/pointwise_pi.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2021 Alex J. Best. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex J. Best
+-/
+import algebra.pointwise
+import algebra.module.pi
+
+/-!
+# Pointwise actions on sets in Pi types
+
+This file contains lemmas about pointwise actions on sets in Pi types.
+
+## Tags
+
+set multiplication, set addition, pointwise addition, pointwise multiplication, pi
+
+-/
+
+open_locale pointwise
+open set
+
+@[to_additive]
+lemma smul_univ_pi {K R : Type*} [has_scalar K R] (ι : Type*) {r : K} (t : ι → set R) :
+  r • pi (univ : set ι) t = pi (univ : set ι) (λ (i : ι), r • t i) :=
+begin
+  ext x,
+  simp only [mem_smul_set, mem_univ_pi],
+  split; intro h,
+  { rcases h with ⟨h_w, h_h_left, rfl⟩,
+    simp only [pi.smul_apply],
+    intro i,
+    refine ⟨h_w i, h_h_left i, rfl⟩, },
+  { use (λ i, classical.some (h i)),
+    split,
+    { exact λ i, (classical.some_spec (h i)).left, },
+    { ext i,
+      exact (classical.some_spec (h i)).right, }, },
+end
+
+lemma smul_pi' {K R : Type*} [group_with_zero K] [mul_action K R] {ι : Type*} {r : K}
+  (t : ι → set R) (S : set ι) (hr : r ≠ 0) : r • S.pi t = S.pi (λ (i : ι), r • t i) :=
+begin
+  ext x,
+  simp only [mem_smul_set, mem_pi],
+  split; intro h,
+  { rcases h with ⟨h_w, h_h_left, rfl⟩,
+    simp only [mul_eq_mul_left_iff, pi.smul_apply],
+    intros i hi,
+    refine ⟨h_w i, h_h_left i hi, rfl⟩, },
+  classical,
+  { existsi (λ i, if hiS : i ∈ S then classical.some (h i hiS) else r⁻¹ • x i),
+    dsimp,
+    split,
+    { intros i hiS,
+      split_ifs,
+      exact (classical.some_spec (h i hiS)).left, },
+    { ext i,
+      rw [pi.smul_apply],
+      split_ifs with hiS,
+      exact (classical.some_spec (h i hiS)).right,
+      rw [smul_smul, mul_inv_cancel hr, one_smul], }, },
+end
+
+@[to_additive]
+lemma smul_pi {K R : Type*} [group K] [mul_action K R] {ι : Type*} {r : K} (t : ι → set R)
+  (S : set ι) : r • S.pi t = S.pi (λ (i : ι), r • t i) :=
+begin
+  ext x,
+  simp only [mem_smul_set, mem_pi],
+  split; intro h,
+  { rcases h with ⟨h_w, h_h_left, rfl⟩,
+    simp only [mul_eq_mul_left_iff, pi.smul_apply],
+    intros i hi,
+    refine ⟨h_w i, h_h_left i hi, rfl⟩, },
+  classical,
+  { existsi (λ i, if hiS : i ∈ S then classical.some (h i hiS) else r⁻¹ • x i),
+    dsimp,
+    split,
+    { intros i hiS,
+      split_ifs,
+      exact (classical.some_spec (h i hiS)).left, },
+    { ext i,
+      rw [pi.smul_apply],
+      split_ifs with hiS,
+      exact (classical.some_spec (h i hiS)).right,
+      rw [smul_smul, mul_right_inv, one_smul], }, },
+end
```

Comment:
Golfed using the `smul_pi_subset` lemma above:
```suggestion
subset.antisymm (smul_pi_subset _ _) $ λ x h,
  ⟨λ i, r⁻¹ • x i, λ i hiS, mem_smul_set_iff_inv_smul_mem.mp (h i hiS), smul_inv_smul _ _⟩
```
The use of `classical.some` here is a distraction.

Reply:
Oops yes, the group with zero version is the one I actually needed, I just added this in case it was useful but my proof is very excessive ahah

Reply:
Sure, but you can prove the group_with_zero version in terms of the group version as `smul_pi (units.mk0 r r0) S r` or whatever the right argument order is.